src/block_cache/block_mgr.c

```diff
@@ -10,6 +10,86 @@
 
 static void __bm_method_set(WT_BM *, bool);
 
+/*
+ * __bm_close_block_remove --
+ *     Remove a single block handle.
+ */
+static int
+__bm_close_block_remove(WT_SESSION_IMPL *session, WT_BLOCK *block)
+{
+    u_int i;
+
+    /* Discard any references we're holding. */
+    for (i = 0; i < block->related_next; ++i) {
+        --block->related[i]->ref;
+        block->related[i] = NULL;
+    }
+
+    /* Discard the block structure. */
+    return (__wt_block_close(session, block));
+}
+
+/*
+ * __bm_close_block --
+ *     Close a single block handle, removing the handle if it's no longer useful.
+ */
+static int
+__bm_close_block(WT_SESSION_IMPL *session, WT_BLOCK *block)
+{
+    WT_CONNECTION_IMPL *conn;
+    WT_DECL_RET;
+    bool found;
+
+    conn = S2C(session);
+
+    __wt_verbose(session, WT_VERB_BLKCACHE, "close: %s", block->name);
+
+    __wt_spin_lock(session, &conn->block_lock);
+    if (block->ref > 0 && --block->ref > 0) {
+        __wt_spin_unlock(session, &conn->block_lock);
+        return (0);
+    }
+
+    /* You can't close files during a checkpoint. */
+    WT_ASSERT(
+      session, block->ckpt_state == WT_CKPT_NONE || block->ckpt_state == WT_CKPT_PANIC_ON_FAILURE);
+
+    /*
+     * Every time we remove a block, we may have sufficiently decremented other references to allow
+     * other blocks to be removed. It's unlikely for blocks to reference each other but it's not out
+     * of the question, either. Loop until we don't find anything to close.
+     */
+    do {
+        found = false;
+        TAILQ_FOREACH (block, &conn->blockqh, q)
+            if (block->ref == 0) {
+                found = true;
+                WT_TRET(__bm_close_block_remove(session, block));
+                break;
+            }
+    } while (found);
+    __wt_spin_unlock(session, &conn->block_lock);
+
+    return (ret);
+}
+
+/*
+ * __bm_drain_and_close --
+ *     Drain the list of waiting WT_BLOCK structures.
+ */
+static int
+__bm_drain_and_close(WT_SESSION_IMPL *session, WT_BLOCK *block)
+{
+    if (block == NULL)
+        return (0);
+
+    /* Close any WT_BLOCK structures queued to be drained, from the end to the beginning. */
+    WT_RET(__bm_drain_and_close(session, block->ckpt_drain));
```

Comment:
This function is slightly confusing because we're draining a list of block structures, while the arg is a single block structure. Maybe a bit more comment about what the list of block structures means to this block would be helpful. Also the phrase `from the end to the beginning` is misplaced in that comment as it presumably refers to the implementation in the function it's calling.

Edit later: Ok, this is a recursive function. I think that should be called out. My eyes did not register the function header name and the fucntion called as the same. It is hinted at with `from the end to the beginning` but it is worth clearly noting.

Reply:
Fixed in 1a0093c59.